File: 0871-keys-and-rooms/0871-keys-and-rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Set<Integer> seen = new HashSet<>();
        Stack<Integer> stack = new Stack<>();
        stack.add(0);
        seen.add(0);
        while (!stack.isEmpty()) {
            int node = stack.pop();
            for (int nb : rooms.get(node)) {
                if (!seen.contains(nb)) {
                    seen.add(nb);
                    stack.add(nb);
                }
            }
        }
        return rooms.size() == seen.size();
    }
}

File: 0871-keys-and-rooms/0871-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        seen = {0}
        stack = [0]
        while stack:
            node = stack.pop()
            for nb in rooms[node]:
                if nb not in seen:
                    seen.add(nb)
                    stack.append(nb)
        return len(seen) == len(rooms)
